File: Backend/app/api/payloads/media.py
from __future__ import annotations

from pathlib import Path
from typing import Any

from flask import current_app, url_for

from ..common.validation import derive_orientation_from_dimensions


def db():
    return current_app.extensions["db"]


def worker():
    return current_app.extensions["job_worker"]


def image_payload(image: dict[str, Any]) -> dict[str, Any]:
    is_stored = image.get("status") in {"stored", "deleted"}
    is_video = image.get("media_type") == "video"
    source_image_id = str(image.get("source_image_id") or "").strip() or None
    source_image_exists = False
    if source_image_id:
        source_image_exists = db().get_image(source_image_id) is not None
    poster_url = (
        url_for("api.get_poster_file", image_id=image["id"], _external=True)
        if is_stored and is_video
        else None
    )
    thumbnail_url = (
        url_for(
            "api.get_thumbnail_file",
            image_id=image["id"],
            _external=True,
            v="video-square-v1" if is_video else None,
        )
        if is_stored
        else None
    )
    is_imported_image = str(image.get("model_id") or "").strip() == "Imported"
    return {
        "id": image["id"],
        "status": image["status"],
        "media_type": image.get("media_type", "image"),
        "mime_type": image["mime_type"],
        "width": image["width"],
        "height": image["height"],
        "prompt": "" if is_imported_image else image["prompt"],
        "default_positive_prompt": (
            "" if is_imported_image else image["default_positive_prompt"]
        ),
        "default_negative_prompt": (
            "" if is_imported_image else image["default_negative_prompt"]
        ),
        "final_positive_prompt": (
            "" if is_imported_image else image["final_positive_prompt"]
        ),
        "model_id": image["model_id"],
        "loras": image["loras"],
        "tags": image.get("tags", []),
        "caption": str(image.get("caption") or ""),
        "num_inference_steps": int(image.get("num_inference_steps", 40)),
        "guidance_scale": float(image.get("guidance_scale", 5.0)),
        "image_orientation": derive_orientation_from_dimensions(
            image.get("width"),
            image.get("height"),
            fallback=str(image.get("image_orientation", "landscape")),
        ),
        "source_image_id": source_image_id,
        "source_image_exists": source_image_exists,
        "scale_factor": image["scale_factor"],
        "is_upscaled": bool(image["is_upscaled"]),
        "duration_seconds": image.get("duration_seconds"),
        "generation_duration_seconds": image.get("generation_duration_seconds"),
        "fps": image.get("fps"),
        "num_frames": image.get("num_frames"),
        "rating": int(image.get("rating", 0)),
        "created_at": image["created_at"],
        "stored_at": image["stored_at"],
        "deleted_at": image.get("deleted_at"),
        "file_url": url_for("api.get_file", image_id=image["id"], _external=True),
        "poster_url": poster_url,
        "thumbnail_url": thumbnail_url,
        "preview_url": thumbnail_url
        or poster_url
        or url_for("api.get_file", image_id=image["id"], _external=True),
    }


def is_still_media(image: dict[str, Any]) -> bool:
    return image.get("media_type", "image") in {"image", "gif"}


def ensure_video_poster(image: dict[str, Any]) -> dict[str, Any]:
    if image.get("media_type") != "video" or image.get("status") != "stored":
        return image
    poster_path_value = str(image.get("poster_path") or "").strip()
    if poster_path_value and Path(poster_path_value).exists():
        return image

    generator = getattr(worker(), "generator", None)
    if generator is None or not hasattr(generator, "create_video_poster"):
        return image

    source_image_path = _source_image_path_for_poster(image)
    try:
        poster_path, poster_mime_type = generator.create_video_poster(
            video_path=Path(image["file_path"]),
            output_dir=Path(current_app.config["TEMP_DIR"]),
            image_id=str(image["id"]),
            fallback_source_image_path=source_image_path,
            width=int(image["width"]),
            height=int(image["height"]),
        )
        if poster_path is None:
            return image
        return (
            db().update_image_poster(
                str(image["id"]),
                poster_path=poster_path,
                poster_mime_type=poster_mime_type,
            )
            or image
        )
    except Exception:
        current_app.logger.warning(
            "Failed to ensure poster for video %s",
            image.get("id"),
            exc_info=True,
        )
        return image


def _source_image_path_for_poster(image: dict[str, Any]) -> Path | None:
    source_image_id = str(image.get("source_image_id") or "").strip()
    if not source_image_id:
        return None
    source_image = db().get_image(source_image_id)
    if source_image is None:
        return None
    return Path(source_image["file_path"])

File: App/lib/models/media.dart
import 'generation.dart';

class LocalImageSource {
  LocalImageSource({required this.path, required this.name});

  final String path;
  final String name;
}

class ImageRecord {
  ImageRecord({
    required this.id,
    required this.status,
    required this.mediaType,
    required this.mimeType,
    required this.width,
    required this.height,
    required this.prompt,
    required this.defaultPositivePrompt,
    required this.defaultNegativePrompt,
    required this.finalPositivePrompt,
    required this.modelId,
    required this.loras,
    required this.tags,
    required this.caption,
    required this.numInferenceSteps,
    required this.guidanceScale,
    required this.imageOrientation,
    required this.fileUrl,
    required this.previewUrl,
    this.thumbnailUrl,
    required this.createdAt,
    required this.rating,
    this.storedAt,
    this.deletedAt,
    this.sourceImageId,
    this.sourceImageExists = false,
    this.scaleFactor,
    this.isUpscaled = false,
    this.posterUrl,
    this.durationSeconds,
    this.generationDurationSeconds,
    this.fps,
    this.numFrames,
  });

  final String id;
  final String status;
  final String mediaType;
  final String mimeType;
  final int width;
  final int height;
  final String prompt;
  final String defaultPositivePrompt;
  final String defaultNegativePrompt;
  final String finalPositivePrompt;
  final String modelId;
  final List<SelectedLora> loras;
  final List<String> tags;
  final String caption;
  final int numInferenceSteps;
  final double guidanceScale;
  final String imageOrientation;
  final String fileUrl;
  final String previewUrl;
  final String? thumbnailUrl;
  final String createdAt;
  final int rating;
  final String? storedAt;
  final String? deletedAt;
  final String? sourceImageId;
  final bool sourceImageExists;
  final double? scaleFactor;
  final bool isUpscaled;
  final String? posterUrl;
  final double? durationSeconds;
  final double? generationDurationSeconds;
  final int? fps;
  final int? numFrames;

  bool get isPending => status == 'queued' || status == 'running';
  bool get isReady => status == 'stored';
  bool get isDeleted => status == 'deleted';
  bool get isVideo => mediaType == 'video';
  bool get isGif => mediaType == 'gif' || mimeType == 'image/gif';
  bool get isImage => !isVideo && !isGif;
  bool get isStillImage => isImage || isGif;
  bool get canPreviewFile => isReady || isDeleted;
  bool get canDownload => isReady || isDeleted;
  bool get canEditMetadata => isReady;
  bool get canScale => (isReady || isPending) && (isVideo || isImage);
  bool get canRegenerate => isReady && (isVideo || isImage);

  String get mediaTypeLabel => isVideo
      ? 'Video'
      : isGif
      ? 'GIF'
      : 'Image';

  ImageRecord copyWith({
    String? status,
    List<String>? tags,
    String? caption,
    int? rating,
    String? storedAt,
    String? fileUrl,
    bool? sourceImageExists,
    String? previewUrl,
    String? thumbnailUrl,
    String? posterUrl,
  }) {
    return ImageRecord(
      id: id,
      status: status ?? this.status,
      mediaType: mediaType,
      mimeType: mimeType,
      width: width,
      height: height,
      prompt: prompt,
      defaultPositivePrompt: defaultPositivePrompt,
      defaultNegativePrompt: defaultNegativePrompt,
      finalPositivePrompt: finalPositivePrompt,
      modelId: modelId,
      loras: loras,
      tags: tags ?? this.tags,
      caption: caption ?? this.caption,
      numInferenceSteps: numInferenceSteps,
      guidanceScale: guidanceScale,
      imageOrientation: imageOrientation,
      fileUrl: fileUrl ?? this.fileUrl,
      previewUrl: previewUrl ?? this.previewUrl,
      thumbnailUrl: thumbnailUrl ?? this.thumbnailUrl,
      createdAt: createdAt,
      rating: rating ?? this.rating,
      storedAt: storedAt ?? this.storedAt,
      deletedAt: deletedAt,
      sourceImageId: sourceImageId,
      sourceImageExists: sourceImageExists ?? this.sourceImageExists,
      scaleFactor: scaleFactor,
      isUpscaled: isUpscaled,
      posterUrl: posterUrl ?? this.posterUrl,
      durationSeconds: durationSeconds,
      generationDurationSeconds: generationDurationSeconds,
      fps: fps,
      numFrames: numFrames,
    );
  }

  factory ImageRecord.fromJson(Map<String, dynamic> json) {
    return ImageRecord(
      id: json['id'] as String,
      status: json['status'] as String,
      mediaType: json['media_type'] as String? ?? 'image',
      mimeType: json['mime_type'] as String? ?? 'image/png',
      width: json['width'] as int,
      height: json['height'] as int,
      prompt: json['prompt'] as String,
      defaultPositivePrompt: json['default_positive_prompt'] as String,
      defaultNegativePrompt: json['default_negative_prompt'] as String,
      finalPositivePrompt: json['final_positive_prompt'] as String,
      modelId: json['model_id'] as String,
      loras: ((json['loras'] as List<dynamic>? ?? <dynamic>[])
          .cast<Map<String, dynamic>>()
          .map(SelectedLora.fromJson)
          .toList()),
      tags: (json['tags'] as List<dynamic>? ?? <dynamic>[])
          .map((item) => item as String)
          .toList(),
      caption: json['caption'] as String? ?? '',
      numInferenceSteps: (json['num_inference_steps'] as num?)?.toInt() ?? 40,
      guidanceScale: (json['guidance_scale'] as num?)?.toDouble() ?? 5.0,
      imageOrientation: json['image_orientation'] as String? ?? 'landscape',
      fileUrl: json['file_url'] as String,
      previewUrl: json['preview_url'] as String? ?? json['file_url'] as String,
      thumbnailUrl: json['thumbnail_url'] as String?,
      createdAt: json['created_at'] as String,
      rating: (json['rating'] as num?)?.toInt() ?? 0,
      storedAt: json['stored_at'] as String?,
      deletedAt: json['deleted_at'] as String?,
      sourceImageId: json['source_image_id'] as String?,
      sourceImageExists:
          json['source_image_exists'] as bool? ??
          ((json['source_image_id'] as String?)?.trim().isNotEmpty ?? false),
      scaleFactor: (json['scale_factor'] as num?)?.toDouble(),
      isUpscaled: json['is_upscaled'] as bool? ?? false,
      posterUrl: json['poster_url'] as String?,
      durationSeconds: (json['duration_seconds'] as num?)?.toDouble(),
      generationDurationSeconds: (json['generation_duration_seconds'] as num?)
          ?.toDouble(),
      fps: (json['fps'] as num?)?.toInt(),
      numFrames: (json['num_frames'] as num?)?.toInt(),
    );
  }
}
